src/core/api.rs

```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
-        access_policy: AccessPolicy,
-    ) -> Result<(SymmetricKey<SEED_LENGTH>, Encapsulation), Error> {
-        encaps(
+        ap: &str,
+    ) -> Result<UserSecretKey, Error> {
+        let ap = &AccessPolicy::parse(ap)?;
+        usk_keygen(
             &mut *self.rng.lock().expect("Mutex lock failed!"),
-            pk,
-            &policy.generate_point_coordinates(access_policy)?,
+            msk,
+            policy.generate_semantic_space_coordinates(ap.clone())?,
         )
     }
+}
 
-    /// Attempts opening the given `Covercrypt` encapsulation. Returns the
-    /// encapsulated key upon success and `None` otherwise.
-    // TODO document error cases.
-    pub fn decaps(
-        &self,
-        usk: &UserSecretKey,
-        encapsulation: &Encapsulation,
-    ) -> Result<Option<SymmetricKey<SEED_LENGTH>>, Error> {
-        decaps(usk, encapsulation)
-    }
+/// Authenticated Encryption trait
+pub trait AE<const KEY_LENGTH: usize, const NONCE_LENGTH: usize, const MAC_LENGTH: usize> {
+    const KEY_LENGTH: usize;
+    const NONCE_LENGTH: usize;
+    const MAC_LENGTH: usize;
+    fn encrypt(
+        key: &SymmetricKey<KEY_LENGTH>,
+        ptx: &[u8],
+        ad: Option<&[u8]>,
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<Vec<u8>, Error>;
 
-    /// Encrypts the given plaintext using the given symmetric key.
-    ///
-    /// The encryption scheme used is AES-256 GCM.
-    ///
-    /// - `symmetric_key`   : AES key
-    /// - `plaintext`       : data to be encrypted
-    /// - `ad`              : optional associated data
-    pub fn dem_encrypt(
-        &self,
-        symmetric_key: &SymmetricKey<SEED_LENGTH>,
-        plaintext: &[u8],
+    fn decrypt(key: &Secret<KEY_LENGTH>, ctx: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, Error>;
+}
+
+impl AE<{ Self::KEY_LENGTH }, { Self::NONCE_LENGTH }, { Self::MAC_LENGTH }> for Aes256Gcm {
+    const KEY_LENGTH: usize = Aes256Gcm::KEY_LENGTH;
+    const MAC_LENGTH: usize = Aes256Gcm::MAC_LENGTH;
+    const NONCE_LENGTH: usize = Aes256Gcm::NONCE_LENGTH;
+    fn encrypt(
+        key: &SymmetricKey<{ Self::KEY_LENGTH }>,
+        ptx: &[u8],
         ad: Option<&[u8]>,
+        rng: &mut impl CryptoRngCore,
     ) -> Result<Vec<u8>, Error> {
-        let aes256gcm = Aes256Gcm::new(symmetric_key);
-        let nonce = Nonce::new(&mut *self.rng.lock().expect("could not lock mutex"));
-        let mut ciphertext = aes256gcm.encrypt(&nonce, plaintext, ad)?;
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
+        let nonce = &self::Nonce::<{ Self::NONCE_LENGTH }>::new(&mut *rng);
+        let aes = Self::new(key);
+        let mut ciphertext = aes
+            .encrypt(nonce, ptx, ad)
+            .map_err(Error::CryptoCoreError)?;
+        let mut res = Vec::with_capacity(ptx.len() + Self::MAC_LENGTH + Self::NONCE_LENGTH);
+        res.extend(&Nonce::to_bytes(nonce));
```

Comment:
This should work and avoids a copy:
```suggestion
        res.extend(&nonce);
```

Reply:
extend needs an iterable type. Nonce does not have the trait implemented.

Reply:
OK, my bad, Idk why but `Deref` is not implemented for this type. Then you should use `as_bytes()` instead of `to_bytes()` to avoid a copy.